shell/platform/embedder/embedder.cc

```diff
@@ -1249,6 +1249,158 @@ void PopulateAOTSnapshotMappingCallbacks(
   }
 }
 
+// Translates engine semantic nodes to embedder semantic nodes.
+FlutterSemanticsNode CreateEmbedderSemanticsNode(
+    const flutter::SemanticsNode& node) {
+  SkMatrix transform = node.transform.asM33();
+  FlutterTransformation flutter_transform{
+      transform.get(SkMatrix::kMScaleX), transform.get(SkMatrix::kMSkewX),
+      transform.get(SkMatrix::kMTransX), transform.get(SkMatrix::kMSkewY),
+      transform.get(SkMatrix::kMScaleY), transform.get(SkMatrix::kMTransY),
+      transform.get(SkMatrix::kMPersp0), transform.get(SkMatrix::kMPersp1),
+      transform.get(SkMatrix::kMPersp2)};
+  return {
+      sizeof(FlutterSemanticsNode),
+      node.id,
+      static_cast<FlutterSemanticsFlag>(node.flags),
+      static_cast<FlutterSemanticsAction>(node.actions),
+      node.textSelectionBase,
+      node.textSelectionExtent,
+      node.scrollChildren,
+      node.scrollIndex,
+      node.scrollPosition,
+      node.scrollExtentMax,
+      node.scrollExtentMin,
+      node.elevation,
+      node.thickness,
+      node.label.c_str(),
+      node.hint.c_str(),
+      node.value.c_str(),
+      node.increasedValue.c_str(),
+      node.decreasedValue.c_str(),
+      static_cast<FlutterTextDirection>(node.textDirection),
+      FlutterRect{node.rect.fLeft, node.rect.fTop, node.rect.fRight,
+                  node.rect.fBottom},
+      flutter_transform,
+      node.childrenInTraversalOrder.size(),
+      node.childrenInTraversalOrder.data(),
+      node.childrenInHitTestOrder.data(),
+      node.customAccessibilityActions.size(),
+      node.customAccessibilityActions.data(),
+      node.platformViewId,
+  };
+}
+
+// Translates engine semantic custom actions to embedder semantic custom
+// actions.
+FlutterSemanticsCustomAction CreateEmbedderSemanticsCustomAction(
+    const flutter::CustomAccessibilityAction& action) {
+  return {
+      sizeof(FlutterSemanticsCustomAction),
+      action.id,
+      static_cast<FlutterSemanticsAction>(action.overrideId),
+      action.label.c_str(),
+      action.hint.c_str(),
+  };
+}
+
+// Creates a callback that receives semantic updates from the engine
+// and notifies the embedder's callback(s). Returns null if the embedder
+// did not register any callbacks.
+flutter::PlatformViewEmbedder::UpdateSemanticsCallback
+CreateEmbedderSemanticsUpdateCallback(const FlutterProjectArgs* args,
+                                      void* user_data) {
+  // The embedder can register the new callback, or the old callbacks, or
+  // nothing at all. Handle the case where the embedder registered the 'new'
+  // callback.
+  if (SAFE_ACCESS(args, update_semantics_callback, nullptr) != nullptr) {
+    return [ptr = args->update_semantics_callback, user_data](
```

Comment:
Can you return a named function object here rather than a callback? I think this lambda is long enough that it might benefit from a name.

Reply:
I split the callbacks up into separate methods in this commit: https://github.com/flutter/engine/pull/37129/commits/b3f4f8261b1b2b67a3280bce28b7fe97c8e2c35d. Let me know what you think!

FYI I do need to return a lambda since it's what the `PlatformViewEmbedder` expects, and, we need to capture values.

Reply:
Yeah, that's better 👍

Reply:
Sweet, marking this as resolved